HTTPDownloader fails to install new engine

Attempt to create a new project in IDE fails.

Regression introduced by #8953

Logs from project-manager:
```
Enso Project Manager
Version:    0.0.0-dev
Built with: scala-2.13.11 for GraalVM 21.0.2
Built from: wip/akirathan/fix-downloader @ 97033a2ff4367ff2191f777d5f03b5eaaf379734
Built on:   Linux (amd64)

[DEBUG] [2024-02-29T11:39:53+01:00] [org.enso.projectmanager.infrastructure.languageserver.LanguageServerRegistry] started (org.enso.projectmanager.infrastructure.languageserver.LanguageServerRegistry@6807b98d)
[DEBUG] [2024-02-29T11:39:53+01:00] [org.enso.projectmanager.infrastructure.languageserver.ShutdownHookActivator] started (org.enso.projectmanager.infrastructure.languageserver.ShutdownHookActivator@5456bf08)
[DEBUG] [2024-02-29T11:39:53+01:00] [org.enso.projectmanager.infrastructure.languageserver.ShutdownHookActivator] now watched by Actor[akka://project-manager/system/eventStreamUnsubscriber-1#1273140297]
[INFO] [2024-02-29T11:39:53+01:00] [org.enso.projectmanager.boot.ProjectManager$] Started server at 127.0.0.1:30535, press enter to kill server
[DEBUG] [2024-02-29T11:39:57+01:00] [org.enso.jsonrpc.MessageHandlerSupervisor] started (org.enso.jsonrpc.MessageHandlerSupervisor@52a9291a)
[DEBUG] [2024-02-29T11:39:57+01:00] [org.enso.jsonrpc.MessageHandler] started (org.enso.jsonrpc.MessageHandler@30ee003a)
[DEBUG] [2024-02-29T11:39:57+01:00] [org.enso.jsonrpc.MessageHandlerSupervisor] now supervising Actor[akka://project-manager/user/message-handler-supervisor-c3a6b096-be99-479b-b157-10cb09f7e724/message-handler-69f9bc77-63f0-482a-a64f-3ca0f6223a6f#-116489347]
[INFO] [2024-02-29T11:39:57+01:00] [org.enso.projectmanager.protocol.ClientController] Client connected to Project Manager [69f9bc77-63f0-482a-a64f-3ca0f6223a6f]
[DEBUG] [2024-02-29T11:39:57+01:00] [org.enso.projectmanager.protocol.ClientController] started (org.enso.projectmanager.protocol.ClientController@7afc6a86)
[DEBUG] [2024-02-29T11:39:57+01:00] [org.enso.jsonrpc.MessageHandlerSupervisor] now watched by Actor[akka://project-manager/system/Materializers/StreamSupervisor-0/$$d#-1850246566]
[DEBUG] [2024-02-29T11:39:57+01:00] [org.enso.projectmanager.protocol.ClientController] now supervising Actor[akka://project-manager/user/manager-client-controller-69f9bc77-63f0-482a-a64f-3ca0f6223a6f/request-handler-ProjectList-a2888a28-895e-44f0-8bc0-2c94ff3cf26a#1352221172]
[DEBUG] [2024-02-29T11:39:57+01:00] [org.enso.projectmanager.protocol.ClientController] now supervising Actor[akka://project-manager/user/manager-client-controller-69f9bc77-63f0-482a-a64f-3ca0f6223a6f/request-handler-ProjectList-ed3701d7-d64d-49ff-a5c5-17849c5e4bb7#-2118321855]
[DEBUG] [2024-02-29T11:39:57+01:00] [org.enso.projectmanager.requesthandler.ProjectListHandler] started (org.enso.projectmanager.requesthandler.ProjectListHandler@65e5e8e8)
[DEBUG] [2024-02-29T11:39:57+01:00] [org.enso.projectmanager.requesthandler.ProjectListHandler] started (org.enso.projectmanager.requesthandler.ProjectListHandler@d93a0fe)
[DEBUG] [2024-02-29T11:39:57+01:00] [org.enso.projectmanager.requesthandler.ProjectListHandler] stopped
[DEBUG] [2024-02-29T11:39:57+01:00] [org.enso.projectmanager.requesthandler.ProjectListHandler] stopped
[DEBUG] [2024-02-29T11:40:00+01:00] [org.enso.projectmanager.protocol.ClientController] now supervising Actor[akka://project-manager/user/manager-client-controller-69f9bc77-63f0-482a-a64f-3ca0f6223a6f/request-handler-ProjectCreate-c4c9594f-1fd2-4526-98b5-fa939af7088b#290794730]
[DEBUG] [2024-02-29T11:40:00+01:00] [org.enso.projectmanager.requesthandler.ProjectCreateHandler] started (org.enso.projectmanager.requesthandler.ProjectCreateHandler@58b14ee2)
[DEBUG] [2024-02-29T11:40:00+01:00] [org.enso.projectmanager.infrastructure.log.Slf4jLogging] Creating project [New Project 2, e481f00c-b28d-4ffa-be12-0294d78ca25d, None].
[INFO] [2024-02-29T11:40:00+01:00] [org.enso.projectmanager.infrastructure.log.Slf4jLogging] Created project with actual name [New Project 2].
[DEBUG] [2024-02-29T11:40:00+01:00] [org.enso.projectmanager.infrastructure.log.Slf4jLogging] Project name [New Project 2] validated by [org.enso.projectmanager.service.validation.ProjectNameValidator@410bba2a].
[DEBUG] [2024-02-29T11:40:00+01:00] [org.enso.projectmanager.infrastructure.log.Slf4jLogging] Checked if the project name [New Project 2] exists in [org.enso.projectmanager.infrastructure.repository.ProjectFileRepository@3107a596].
[DEBUG] [2024-02-29T11:40:00+01:00] [org.enso.projectmanager.infrastructure.log.Slf4jLogging] Found a path [/home/pavel/enso/projects/NewProject2] for a new project [New Project 2, e481f00c-b28d-4ffa-be12-0294d78ca25d].
[WARN] [2024-02-29T11:40:00+01:00] [org.enso.runtimeversionmanager.components.RuntimeVersionManager] Ignoring the minimum required engine version check [2023.2.1-nightly.2023.11.2] for the development version [0.0.0-dev].
[INFO] [2024-02-29T11:40:00+01:00] [org.enso.runtimeversionmanager.components.RuntimeVersionManager] The engine [0.0.0-dev] found.
[INFO] [2024-02-29T11:40:00+01:00] [org.enso.runtimeversionmanager.components.RuntimeVersionManager] GraalVM runtime [GraalVM 23.1.2 Java 21.0.2] not found.
[INFO] [2024-02-29T11:40:01+01:00] [org.enso.runtimeversionmanager.components.RuntimeVersionManager] Installing GraalVM runtime [GraalVM 23.1.2 Java 21.0.2].
[DEBUG] [2024-02-29T11:40:01+01:00] [org.enso.downloader.http.HTTPDownload] Fetching [https://api.github.com/repos/graalvm/graalvm-ce-builds/releases/tags/jdk-21.0.2].
[DEBUG] [2024-02-29T11:40:01+01:00] [org.enso.downloader.http.HTTPDownload] Downloading [https://github.com/graalvm/graalvm-ce-builds/releases/download/jdk-21.0.2/graalvm-community-jdk-21.0.2_linux-x64_bin.tar.gz] to [/tmp/enso-install-runtime9055653635292685988/graalvm-community-jdk-21.0.2_linux-x64_bin.tar.gz].
[DEBUG] [2024-02-29T11:40:01+01:00] [org.enso.runtimeversionmanager.components.RuntimeVersionManager] Downloading [graalvm-community-jdk-21.0.2_linux-x64_bin.tar.gz].
[DEBUG] [2024-02-29T11:40:02+01:00] [org.enso.runtimeversionmanager.components.RuntimeVersionManager] Extracting [/tmp/enso-install-runtime9055653635292685988/graalvm-community-jdk-21.0.2_linux-x64_bin.tar.gz].
[DEBUG] [2024-02-29T11:40:02+01:00] [org.enso.downloader.archive.Archive] Opening [/tmp/enso-install-runtime9055653635292685988/graalvm-community-jdk-21.0.2_linux-x64_bin.tar.gz].
[ERROR] [2024-02-29T11:40:02+01:00] [org.enso.projectmanager.requesthandler.ProjectCreateHandler] Request ProjectCreate with Number(2) failed due to ProjectCreateFailed(null).
[DEBUG] [2024-02-29T11:40:02+01:00] [org.enso.projectmanager.requesthandler.ProjectCreateHandler] stopped
[WARN] [2024-02-29T11:40:02+01:00] [org.enso.jsonrpc.MessageHandlerSupervisor] Resuming after error.
java.lang.NullPointerException: Cannot invoke "String.length()" because "value" is null
	at io.circe.Printer$PrintingFolder.onString(Printer.scala:319)
	at io.circe.Printer$PrintingFolder.onString(Printer.scala:302)
	at io.circe.Json$JString.foldWith(Json.scala:369)
	at io.circe.JsonObject$LinkedHashMapJsonObject.appendToFolder(JsonObject.scala:343)
	at io.circe.Printer$PrintingFolder.onObject(Printer.scala:369)
	at io.circe.Printer$PrintingFolder.onObject(Printer.scala:302)
	at io.circe.Json$JObject.foldWith(Json.scala:431)
	at io.circe.JsonObject$LinkedHashMapJsonObject.appendToFolder(JsonObject.scala:343)
	at io.circe.Printer$PrintingFolder.onObject(Printer.scala:369)
	at io.circe.Printer$PrintingFolder.onObject(Printer.scala:302)
	at io.circe.Json$JObject.foldWith(Json.scala:431)
	at io.circe.Printer.unsafePrintToAppendable(Printer.scala:229)
	at io.circe.Printer.print(Printer.scala:197)
	at io.circe.Json.noSpaces(Json.scala:122)
	at org.enso.jsonrpc.JsonProtocol$.encode(JsonProtocol.scala:166)
	at org.enso.jsonrpc.MessageHandler.org$enso$jsonrpc$MessageHandler$$issueResponseError(MessageHandler.scala:74)
	at org.enso.jsonrpc.MessageHandler$$anonfun$org$enso$jsonrpc$MessageHandler$$established$1.applyOrElse(MessageHandler.scala:49)
	at akka.actor.Actor.aroundReceive(Actor.scala:537)
	at akka.actor.Actor.aroundReceive$(Actor.scala:535)
	at org.enso.jsonrpc.MessageHandler.aroundReceive(MessageHandler.scala:13)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:579)
	at akka.actor.ActorCell.invoke(ActorCell.scala:547)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
	at akka.dispatch.Mailbox.run(Mailbox.scala:231)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
[WARN] [2024-02-29T11:40:02+01:00] [akka.actor.OneForOneStrategy] Cannot invoke "String.length()" because "value" is null
[INFO] [2024-02-29T11:40:02+01:00] [akka.actor.LocalActorRef] Message [org.enso.cli.task.ProgressNotification$TaskUpdate] to Actor[akka://project-manager/user/manager-client-controller-69f9bc77-63f0-482a-a64f-3ca0f6223a6f/request-handler-ProjectCreate-c4c9594f-1fd2-4526-98b5-fa939af7088b#290794730] was not delivered. [1] dead letters encountered, no more dead letters will be logged in next [5.000 min]. If this is not an expected behavior then Actor[akka://project-manager/user/manager-client-controller-69f9bc77-63f0-482a-a64f-3ca0f6223a6f/request-handler-ProjectCreate-c4c9594f-1fd2-4526-98b5-fa939af7088b#290794730] may have terminated unexpectedly. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
```

Logs from IDE:
```
Error creating new project: Input is not in the .gz format
```

## Reproducer
The simplest reproducer to try to download a new engine via launcher:
- `sbt launcher/assembly`
- Make sure there is no engine version `2024.1.1-nightly.2024.2.29` in `java -jar launcher.jar --launcher-log-level trace list`.
  - If there is, remove it from `$HOME/.local/share/enso/dist`
- Try to download the newest engine with: `java -jar launcher.jar --launcher-log-level trace install engine 2024.1.1-nightly.2024.2.29`
This fails with:
```
[INFO] [2024-02-29T11:35:16+01:00] [org.enso.runtimeversionmanager.components.RuntimeVersionManager] The engine [2024.1.1-nightly.2024.2.29] not found.
[INFO] [2024-02-29T11:35:16+01:00] [org.enso.runtimeversionmanager.components.RuntimeVersionManager] Installing the engine [2024.1.1-nightly.2024.2.29].
[DEBUG] [2024-02-29T11:35:16+01:00] [org.enso.downloader.http.HTTPDownload] Fetching [https://api.github.com/repos/enso-org/enso/releases/tags/2024.1.1-nightly.2024.2.29].
[DEBUG] [2024-02-29T11:35:17+01:00] [org.enso.downloader.http.HTTPDownload] Fetching [https://github.com/enso-org/enso/releases/download/2024.1.1-nightly.2024.2.29/manifest.yaml].
[WARN] [2024-02-29T11:35:18+01:00] [org.enso.runtimeversionmanager.components.RuntimeVersionManager] Ignoring the minimum required engine version check [2023.2.1-nightly.2023.11.2] for the development version [0.0.0-dev].
[DEBUG] [2024-02-29T11:35:18+01:00] [org.enso.runtimeversionmanager.components.RuntimeVersionManager] Downloading packages to [/tmp/enso-install7410196257632002202].
[DEBUG] [2024-02-29T11:35:18+01:00] [org.enso.downloader.http.HTTPDownload] Downloading [https://github.com/enso-org/enso/releases/download/2024.1.1-nightly.2024.2.29/enso-engine-2024.1.1-nightly.2024.2.29-linux-amd64.tar.gz] to [/tmp/enso-install7410196257632002202/enso-engine-2024.1.1-nightly.2024.2.29-linux-amd64.tar.gz].
[INFO] [2024-02-29T11:35:18+01:00] [launcher] Downloading enso-engine-2024.1.1-nightly.2024.2.29-linux-amd64.tar.gz.
[INFO] [2024-02-29T11:35:19+01:00] [launcher] Extracting the engine.
[DEBUG] [2024-02-29T11:35:19+01:00] [org.enso.downloader.archive.Archive] Opening [/tmp/enso-install7410196257632002202/enso-engine-2024.1.1-nightly.2024.2.29-linux-amd64.tar.gz].
[ERROR] [2024-02-29T11:35:19+01:00] [org.enso.launcher.cli.Main] A fatal error has occurred: null
[ERROR] [2024-02-29T11:35:19+01:00] [org.enso.launcher.cli.Main] A fatal error has occurred: null
```

--- a/lib/scala/downloader/src/main/java/org/enso/downloader/http/PathProgressBodyHandler.java
+++ b/lib/scala/downloader/src/main/java/org/enso/downloader/http/PathProgressBodyHandler.java
@@ -14,11 +14,15 @@ import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
 import java.util.concurrent.Flow;
 import org.enso.cli.task.TaskProgressImplementation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import scala.Option;
 import scala.Some;
 
 /** A {@link HttpResponse} body handler for {@link Path} that keeps track of the progress. */
 class PathProgressBodyHandler implements HttpResponse.BodyHandler<Path> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PathProgressBodyHandler.class);
+
   private final Path destination;
   private final TaskProgressImplementation<Path> progress;
   private Long total;
@@ -41,6 +45,7 @@ class PathProgressBodyHandler implements HttpResponse.BodyHandler<Path> {
       var reportedLenOpt = responseInfo.headers().firstValueAsLong("Content-Length");
       if (reportedLenOpt.isPresent()) {
         total = reportedLenOpt.getAsLong();
+        LOGGER.debug("Content-Length: {}", total);
       }
     }
     if (total != null) {
@@ -48,6 +53,7 @@ class PathProgressBodyHandler implements HttpResponse.BodyHandler<Path> {
     } else {
       progress.reportProgress(0, Option.empty());
     }
+    LOGGER.debug("Initializing download into {}, estimated total is {}", destination, total);
     WritableByteChannel destChannel;
     try {
       destChannel =
@@ -62,6 +68,7 @@ class PathProgressBodyHandler implements HttpResponse.BodyHandler<Path> {
     private Flow.Subscription subscription;
     private final WritableByteChannel destChannel;
     private final CompletableFuture<Path> result = new CompletableFuture<>();
+    private long downloaded;
 
     ProgressSubscriber(WritableByteChannel destChannel) {
       this.destChannel = destChannel;
@@ -78,7 +85,8 @@ class PathProgressBodyHandler implements HttpResponse.BodyHandler<Path> {
       try {
         for (ByteBuffer item : items) {
           var len = item.remaining();
-          progress.reportProgress(len, total == null ? Option.empty() : Some.apply(total));
+          downloaded += len;
+          progress.reportProgress(downloaded, total == null ? Option.empty() : Some.apply(total));
           destChannel.write(item);
         }
         subscription.request(1);
@@ -90,6 +98,12 @@ class PathProgressBodyHandler implements HttpResponse.BodyHandler<Path> {
 
     @Override
     public void onError(Throwable throwable) {
+      LOGGER.warn(
+          "Error while downloading into {}. Download progress {}/{}",
+          destination,
+          downloaded,
+          total,
+          throwable);
       try {
         destChannel.close();
       } catch (IOException e) {
@@ -100,6 +114,7 @@ class PathProgressBodyHandler implements HttpResponse.BodyHandler<Path> {
 
     @Override
     public void onComplete() {
+      LOGGER.debug("Downloaded complete into {}", destination);
       try {
         destChannel.close();
       } catch (IOException e) {
@@ -110,7 +125,7 @@ class PathProgressBodyHandler implements HttpResponse.BodyHandler<Path> {
 
     @Override
     public CompletionStage<Path> getBody() {
-      return CompletableFuture.completedFuture(destination);
+      return result;
     }
   }
 }

--- a/lib/scala/downloader/src/main/java/org/enso/downloader/http/PathProgressBodyHandler.java
+++ b/lib/scala/downloader/src/main/java/org/enso/downloader/http/PathProgressBodyHandler.java
@@ -86,7 +86,7 @@ class PathProgressBodyHandler implements HttpResponse.BodyHandler<Path> {
         for (ByteBuffer item : items) {
           var len = item.remaining();
           downloaded += len;
-          progress.reportProgress(downloaded, total == null ? Option.empty() : Some.apply(total));
+          progress.reportProgress(downloaded, Option.apply(total));
           destChannel.write(item);
         }
         subscription.request(1);

--- a/lib/scala/downloader/src/main/java/org/enso/downloader/http/StringProgressBodyHandler.java
+++ b/lib/scala/downloader/src/main/java/org/enso/downloader/http/StringProgressBodyHandler.java
@@ -6,7 +6,6 @@ import java.net.http.HttpResponse;
 import java.net.http.HttpResponse.BodySubscriber;
 import java.nio.ByteBuffer;
 import java.nio.charset.Charset;
-import java.nio.file.Path;
 import java.util.List;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
@@ -15,12 +14,13 @@ import org.enso.cli.task.TaskProgressImplementation;
 import scala.Option;
 import scala.Some;
 
-/** A {@link HttpResponse} body handler for {@link Path} that keeps track of the progress. */
+/** A {@link HttpResponse} body handler for {@link String} that keeps track of the progress. */
 public class StringProgressBodyHandler implements HttpResponse.BodyHandler<String> {
   private final ByteArrayOutputStream destination = new ByteArrayOutputStream();
   private final TaskProgressImplementation<?> progress;
   private final Charset encoding;
   private Long total;
+  private long downloaded;
 
   private StringProgressBodyHandler(
       TaskProgressImplementation<?> progress, Charset encoding, Long total) {
@@ -64,7 +64,8 @@ public class StringProgressBodyHandler implements HttpResponse.BodyHandler<Strin
     public void onNext(List<ByteBuffer> items) {
       for (ByteBuffer item : items) {
         var len = item.remaining();
-        progress.reportProgress(len, total == null ? Option.empty() : Some.apply(total));
+        downloaded += len;
+        progress.reportProgress(downloaded, total == null ? Option.empty() : Some.apply(total));
         byte[] bytes = new byte[len];
         item.get(bytes);
         destination.write(bytes, 0, bytes.length);

--- a/lib/scala/downloader/src/main/java/org/enso/downloader/http/StringProgressBodyHandler.java
+++ b/lib/scala/downloader/src/main/java/org/enso/downloader/http/StringProgressBodyHandler.java
@@ -65,7 +65,7 @@ public class StringProgressBodyHandler implements HttpResponse.BodyHandler<Strin
       for (ByteBuffer item : items) {
         var len = item.remaining();
         downloaded += len;
-        progress.reportProgress(downloaded, total == null ? Option.empty() : Some.apply(total));
+        progress.reportProgress(downloaded, Option.apply(total));
         byte[] bytes = new byte[len];
         item.get(bytes);
         destination.write(bytes, 0, bytes.length);

--- a/lib/scala/downloader/src/test/java/org/enso/downloader/http/HttpDownloaderTest.java
+++ b/lib/scala/downloader/src/test/java/org/enso/downloader/http/HttpDownloaderTest.java
@@ -6,6 +6,7 @@ import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.greaterThan;
 import static org.hamcrest.Matchers.lessThan;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.junit.Assert.assertTrue;
 
@@ -13,8 +14,11 @@ import com.sun.net.httpserver.HttpExchange;
 import java.io.IOException;
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Random;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
@@ -97,25 +101,25 @@ public class HttpDownloaderTest {
 
   @Test
   public void downloadBigFileWithProgress() throws Exception {
-    var uriStr = "http://localhost:" + port + "/files/big.tgz";
+    var uriStr = "http://localhost:" + port + "/files/big.txt";
     var uri = new URI(uriStr);
-    Path dest = Files.createTempFile("enso-downloader-test", ".tgz");
-    final int[] progressUpdateCalls = {0};
+    Path dest = Files.createTempFile("enso-downloader-test", ".txt");
+    List<Long> progressCalls = new ArrayList<>();
     final int[] doneCalls = {0};
     var task = HTTPDownload.download(uri, dest);
     var progressListener =
         new ProgressListener<Path>() {
           @Override
           public void progressUpdate(long done, Option<Object> total) {
-            progressUpdateCalls[0]++;
+            progressCalls.add(done);
             assertThat(total.isDefined(), is(true));
             assertThat(total.get(), instanceOf(Long.class));
             long reportedTotal = (Long) total.get();
             assertThat(reportedTotal, is(BigFileHandler.BIG_FILE_SIZE));
             assertThat(
-                "Should send and report just chunk, not the whole file",
+                "Reported progress should not exceed total size",
                 done,
-                lessThan(reportedTotal));
+                lessThanOrEqualTo(reportedTotal));
           }
 
           @Override
@@ -129,8 +133,13 @@ public class HttpDownloaderTest {
     var resp = task.force();
     assertThat(resp.toFile().exists(), is(true));
     assertThat("Done was called exactly once", doneCalls[0], is(1));
-    assertThat(
-        "Progress reported was called at least once", progressUpdateCalls[0], greaterThan(0));
+    assertThat("Progress reported was called at least twice", progressCalls.size(), greaterThan(1));
+    for (int i = 0; i < progressCalls.size() - 1; i++) {
+      var progress = progressCalls.get(i);
+      var nextProgress = progressCalls.get(i + 1);
+      assertThat("Progress should be increasing", progress, lessThan(nextProgress));
+    }
+    BigFileHandler.assertSameFileContent(dest);
     Files.deleteIfExists(dest);
   }
 
@@ -191,33 +200,41 @@ public class HttpDownloaderTest {
   }
 
   private static class BigFileHandler extends SimpleHttpHandler {
-    private static final int CHUNK_SIZE = 1024;
+    private static final int CHUNK_SIZE = 4096;
     private static final long BIG_FILE_SIZE = 10 * CHUNK_SIZE;
-    private static final byte[] BIG_FILE_BYTES = new byte[Math.toIntExact(BIG_FILE_SIZE)];
+    private static final byte[] BIG_FILE_CONTENT;
 
     static {
       var rnd = new Random(42);
-      rnd.nextBytes(BIG_FILE_BYTES);
+      var sb = new StringBuilder();
+      for (int i = 0; i < BIG_FILE_SIZE; i++) {
+        char c = (char) (rnd.nextInt(26) + 'a');
+        sb.append(c);
+      }
+      BIG_FILE_CONTENT = sb.toString().getBytes(StandardCharsets.UTF_8);
     }
 
     @Override
     protected void doHandle(HttpExchange exchange) throws IOException {
-      if (exchange.getRequestURI().toString().equals("/files/big.tgz")) {
+      if (exchange.getRequestURI().toString().equals("/files/big.txt")) {
         long chunks = BIG_FILE_SIZE / CHUNK_SIZE;
         exchange.getResponseHeaders().set("Content-Length", Long.toString(BIG_FILE_SIZE));
-        exchange.getResponseHeaders().set("Content-Type", "application/x-gzip");
-        // Set responseLength to 0 to indicate that the response length is unknown
-        // and force chunking the response.
-        exchange.sendResponseHeaders(200, 0);
+        exchange.getResponseHeaders().set("Content-Type", "text/plain; charset=utf-8");
+        exchange.getResponseHeaders().set("Transfer-Encoding", "chunked");
+        exchange.sendResponseHeaders(200, BIG_FILE_SIZE);
         try (var os = exchange.getResponseBody()) {
           for (int i = 0; i < chunks; i++) {
-            os.write(BIG_FILE_BYTES, i * CHUNK_SIZE, CHUNK_SIZE);
-            os.flush();
+            os.write(BIG_FILE_CONTENT, i * CHUNK_SIZE, CHUNK_SIZE);
           }
         }
       } else {
         exchange.sendResponseHeaders(404, -1);
       }
+    }
+
+    static void assertSameFileContent(Path file) throws IOException {
+      byte[] readbytes = Files.readAllBytes(file);
+      assertThat(readbytes, is(BIG_FILE_CONTENT));
     }
   }
 }

